fs/vfs/fs_dup.c

```diff
@@ -52,27 +52,50 @@
 
 int file_dup(FAR struct file *filep, int minfd, int flags)
 {
-  struct file filep2;
+  FAR struct file *filep2;
   int fd2;
   int ret;
+#ifdef CONFIG_FDSAN
+  uint64_t f_tag_fdsan; /* File owner fdsan tag, init to 0 */
+#endif
 
-  /* Let file_dup3() do the real work */
+#ifdef CONFIG_FDCHECK
+  uint8_t f_tag_fdcheck; /* File owner fdcheck tag, init to 0 */
+#endif
 
-  memset(&filep2, 0, sizeof(filep2));
-  ret = file_dup3(filep, &filep2, flags);
-  if (ret < 0)
+  fd2 = file_allocate(g_root_inode, 0, 0, NULL, minfd, true);
+  if (fd2 < 0)
     {
-      return ret;
+      return fd2;
     }
 
-  fd2 = file_allocate(filep2.f_inode, filep2.f_oflags,
-                      filep2.f_pos, filep2.f_priv, minfd, false);
-  if (fd2 < 0)
+  ret = fs_getfilep(fd2, &filep2);
+#ifdef CONFIG_FDSAN
+  f_tag_fdsan = filep2->f_tag_fdsan;
+#endif
+
+#ifdef CONFIG_FDCHECK
+  f_tag_fdcheck = filep2->f_tag_fdcheck;
+#endif
+  DEBUGASSERT(ret >= 0);
```

Comment:
Better to check this after `fs_getfilep` return before dereferencing `NULL` pointers.

Reply:
> Better to check this after `fs_getfilep` return before dereferencing `NULL` pointers.

@pkarashchenko  We have already determined whether file_allocate can return correctly before this, which should be sufficient, don't you think?

Reply:
fs_getfilep is simple to add the reference count, which should never fail, so DEBUGASSERT is enough here.